src/main/kotlin/exercise4/task2/TransactionalBankAccount.kt

```diff
@@ -80,6 +81,110 @@ import java.util.*
 
 
 
+data class Transaction (
+    val transactionDate: LocalDateTime,
+    val transactionType: Type,
+    val amount : Double,
+    val oldBalance : Double,
+    val newBalance : Double,
+    val transactionStatus : Status
+) {
+    enum class Type {
+        DEPOSIT , WITHDRAWAL
+    }
+
+    enum class Status {
+        SUCCESS, FAILURE
+    }
+}
+
+class TransactionalBankAccount(
+    accountNumber: String,
+    accountHolderName: String,
+    private val transactions: MutableList<Transaction> = mutableListOf()
+) : BankAccount(accountNumber, accountHolderName) {
+
+    override fun deposit(amount: Double) {
+        val oldBalance = getBalance
+        super.deposit(amount)
+        transactions.add(Transaction(currentTime, Transaction.Type.DEPOSIT, amount, oldBalance, getBalance, Transaction.Status.SUCCESS))
+    }
+
+    override fun withdraw(amount: Double): Boolean {
+        val oldBalance = getBalance
+        val result = super.withdraw(amount)
+        val status = if(result) Transaction.Status.SUCCESS else Transaction.Status.FAILURE
+
+        transactions.add(Transaction(currentTime, Transaction.Type.WITHDRAWAL, amount, oldBalance, getBalance, status))
+        return result
+    }
+
+    fun getAllTransactions() : List<Transaction> {
+        return transactions.sortedByDescending { it.transactionDate }
+    }
+
+    fun getAllTransactionsBy(predicate: (Transaction) -> Boolean) : List<Transaction> {
+        return transactions.filter { predicate(it) }.
+                            sortedByDescending { it.transactionDate }
+    }
+
+    fun getTransactionsBetween(startDate: LocalDateTime, endDate: LocalDateTime) : List<Transaction> {
+        return transactions.filter { it.transactionDate in startDate..endDate }.
+                            sortedByDescending { it.transactionDate }
+    }
+
+    fun getAllFailedTransactions() : List<Transaction> {
+        return transactions.filter { it.transactionStatus == Transaction.Status.FAILURE }.
+                            sortedByDescending { it.transactionDate }
+    }
+
+    fun getAllSuccessfulTransactions() : List<Transaction> {
+        return transactions.filter { it.transactionStatus == Transaction.Status.SUCCESS }.
+                            sortedByDescending { it.transactionDate }
+    }
+
+    fun getAllFailedDeposits() : List<Transaction> {
+        return transactions.filter { it.transactionStatus == Transaction.Status.FAILURE && it.transactionType == Transaction.Type.DEPOSIT }.
+                            sortedByDescending { it.transactionDate }
+    }
+
+    fun getAllFailedWithdrawals() : List<Transaction> {
+        return transactions.filter { it.transactionStatus == Transaction.Status.FAILURE && it.transactionType == Transaction.Type.WITHDRAWAL }.
```

Comment:
nit: We could have reused `getAllTransactionsBy` function instead of defining filter and sort every time. :) 
Example: 
```kotlin
fun getAllFailedDeposits(): List<Transaction> = getAllTransactionsBy {
   it.transactionStatus == Transaction.Status.FAILURE && it.transactionType == Transaction.Type.DEPOSIT
}
```